### ISSUE TITLE
A more sensible error message for assertions on arrow functions

### ISSUE THREAD

**Issue by sezanzeb** [microsoft/TypeScript#53450]
# Suggestion

## 🔍 Search Terms

List of keywords you searched for before creating this issue. Write them down here so that others can find this suggestion more easily and help provide feedback.

- 2775
- arrow

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

```ts
const expectFoo = (bar: unknown): asserts bar is Foo => { ... };
```

This line should show a useful error message, like `TSxxxx: Assertions cannot be used on a arrow-functions` 

## 💻 Use Cases

Instead it shows `TS2775: Assertions require every name in the call target to be declared with an explicit type annotation.` as error. And not where the `assert` is, but instead where the function is called.

https://github.com/microsoft/TypeScript/issues/34523#issuecomment-1221400964

> A better error message on this design limitation would have been nice instead of the confusing TS 2775

I also think this is extremely confusing. I just had to look that error up for the third time now, visiting the same StackOverflow post for the third time, and realizing that I had assertions on an arrow function again for the third time.

An error like `TSxxxx: Assertions cannot be used on an arrow-function` would be much better.

**Comment by MartinJohns** [comment]
But arrow-functions can be assertions. They just need an explicit type annotation, as mentioned in TS2775.

**Comment by sezanzeb** [comment]
Ah, this is what I get for only reading the title and then just scrolling all the way down

**Comment by sezanzeb** [comment]
That error pops up where I call that arrow function though, not where I declare it.

**Comment by RyanCavanaugh** [comment]
The error message you're getting is telling you what's wrong

> Assertions require every name in the call target to be declared with an explicit type annotation.

`expectFoo` does not have an explicit type annotation. You can tell because there's no `:` after its name:
```ts
// Not annotated
const expectFoo = ...
// Annotated
const expectFoo: SomeType = ...
```

I'm open to a *correct* rewording of this error message if anyone has one.

**Comment by erfanium** [comment]
Lost one hour of my life because of this confusing error. please re-open the issue
